cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/template/etl/common/DefaultEmitter.java

```diff
@@ -56,6 +58,11 @@ public void emitError(InvalidEntry<T> value) {
     return entryList.iterator();
   }
 
+  public Iterator<InvalidEntry<T>> getErrorIterator() {
+    // we have to return iterator of a copy of list, as we need reset the errorList between stages.
+    return new ArrayList<>(errorList).iterator();
```

Comment:
the normal iterator just updates a single list between each stage of the transform and after all the transform stages are complete, it just has to return the iterator of the list. whereas for our use case, we are interested in error records from each stage, so we have to do a copy. ``runOneIteration`` runs transform on all stages before returning.